File: gesttravaux-java/src/main/java/com/immosync/gesttravaux/models/Bien.java
package com.immosync.gesttravaux.models;

import jakarta.persistence.*;
import java.util.HashSet;
import java.util.Set;

/**
 * Entity representing a property (bien immobilier).
 * Each property belongs to one proprietaire and can have multiple chantiers.
 *
 * @author GestTravaux Pro
 * @version 1.0
 */
@Entity
@Table(name = "BIENS")
public class Bien {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "idBien")
    private Integer idBien;

    @Column(name = "adresseBien", length = 45, nullable = false)
    private String adresseBien;

    @Column(name = "villeBien", length = 45, nullable = false)
    private String villeBien;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "PROPRIETAIRES_idProprietaire", nullable = false)
    private Proprietaire proprietaire;

    @OneToMany(mappedBy = "bien", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Chantier> chantiers = new HashSet<>();

    /**
     * Default constructor required by JPA.
     */
    public Bien() {
    }

    /**
     * Constructor with essential fields.
     *
     * @param adresseBien property address
     * @param villeBien property city
     * @param proprietaire property owner
     */
    public Bien(String adresseBien, String villeBien, Proprietaire proprietaire) {
        this.adresseBien = adresseBien;
        this.villeBien = villeBien;
        this.proprietaire = proprietaire;
    }

    // Getters and Setters

    /**
     * Gets the property ID.
     *
     * @return the property ID
     */
    public Integer getIdBien() {
        return idBien;
    }

    /**
     * Sets the property ID.
     *
     * @param idBien the property ID
     */
    public void setIdBien(Integer idBien) {
        this.idBien = idBien;
    }

    /**
     * Gets the property address.
     *
     * @return the address
     */
    public String getAdresseBien() {
        return adresseBien;
    }

    /**
     * Sets the property address.
     *
     * @param adresseBien the address
     */
    public void setAdresseBien(String adresseBien) {
        this.adresseBien = adresseBien;
    }

    /**
     * Gets the property city.
     *
     * @return the city
     */
    public String getVilleBien() {
        return villeBien;
    }

    /**
     * Sets the property city.
     *
     * @param villeBien the city
     */
    public void setVilleBien(String villeBien) {
        this.villeBien = villeBien;
    }

    /**
     * Gets the property owner.
     *
     * @return the proprietaire
     */
    public Proprietaire getProprietaire() {
        return proprietaire;
    }

    /**
     * Sets the property owner.
     *
     * @param proprietaire the proprietaire
     */
    public void setProprietaire(Proprietaire proprietaire) {
        this.proprietaire = proprietaire;
    }

    /**
     * Gets the chantiers associated with this property.
     *
     * @return set of chantiers
     */
    public Set<Chantier> getChantiers() {
        return chantiers;
    }

    /**
     * Sets the chantiers associated with this property.
     *
     * @param chantiers set of chantiers
     */
    public void setChantiers(Set<Chantier> chantiers) {
        this.chantiers = chantiers;
    }

    /**
     * Adds a chantier to this property.
     *
     * @param chantier the chantier to add
     */
    public void addChantier(Chantier chantier) {
        chantiers.add(chantier);
        chantier.setBien(this);
    }

    /**
     * Removes a chantier from this property.
     *
     * @param chantier the chantier to remove
     */
    public void removeChantier(Chantier chantier) {
        chantiers.remove(chantier);
        chantier.setBien(null);
    }

    /**
     * Gets the full address of the property.
     *
     * @return full address (address, city)
     */
    public String getFullAddress() {
        return adresseBien + ", " + villeBien;
    }

    @Override
    public String toString() {
        return getFullAddress();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Bien)) return false;
        Bien bien = (Bien) o;
        return idBien != null && idBien.equals(bien.getIdBien());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: gesttravaux-symfony/src/Entity/Bien.php
<?php

namespace App\Entity;

use App\Repository\BienRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

/**
 * Bien entity - Properties
 *
 * Represents real estate properties where construction/renovation work takes place
 *
 * @author GestTravaux Pro
 */
#[ORM\Entity(repositoryClass: BienRepository::class)]
#[ORM\Table(name: 'biens')]
class Bien
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idBien')]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    private ?string $adresseBien = null;

    #[ORM\Column(length: 45)]
    private ?string $villeBien = null;

    #[ORM\Column(length: 45, nullable: true)]
    private ?string $fk_idProprietaire = null;

    #[ORM\Column(type: 'decimal', precision: 10, scale: 8, nullable: true)]
    private ?string $latitude = null;

    #[ORM\Column(type: 'decimal', precision: 11, scale: 8, nullable: true)]
    private ?string $longitude = null;

    #[ORM\ManyToOne(targetEntity: Proprietaire::class, inversedBy: 'biens')]
    #[ORM\JoinColumn(name: 'proprietaires_idProprietaire', referencedColumnName: 'idProprietaire', nullable: false)]
    private ?Proprietaire $proprietaire = null;

    /**
     * @var Collection<int, Chantier>
     */
    #[ORM\OneToMany(targetEntity: Chantier::class, mappedBy: 'bien')]
    private Collection $chantiers;

    public function __construct()
    {
        $this->chantiers = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getAdresseBien(): ?string
    {
        return $this->adresseBien;
    }

    public function setAdresseBien(string $adresseBien): static
    {
        $this->adresseBien = $adresseBien;

        return $this;
    }

    public function getVilleBien(): ?string
    {
        return $this->villeBien;
    }

    public function setVilleBien(string $villeBien): static
    {
        $this->villeBien = $villeBien;

        return $this;
    }

    public function getFkIdProprietaire(): ?string
    {
        return $this->fk_idProprietaire;
    }

    public function setFkIdProprietaire(?string $fk_idProprietaire): static
    {
        $this->fk_idProprietaire = $fk_idProprietaire;

        return $this;
    }

    public function getLatitude(): ?string
    {
        return $this->latitude;
    }

    public function setLatitude(?string $latitude): static
    {
        $this->latitude = $latitude;

        return $this;
    }

    public function getLongitude(): ?string
    {
        return $this->longitude;
    }

    public function setLongitude(?string $longitude): static
    {
        $this->longitude = $longitude;

        return $this;
    }

    public function getProprietaire(): ?Proprietaire
    {
        return $this->proprietaire;
    }

    public function setProprietaire(?Proprietaire $proprietaire): static
    {
        $this->proprietaire = $proprietaire;

        return $this;
    }

    /**
     * @return Collection<int, Chantier>
     */
    public function getChantiers(): Collection
    {
        return $this->chantiers;
    }

    public function addChantier(Chantier $chantier): static
    {
        if (!$this->chantiers->contains($chantier)) {
            $this->chantiers->add($chantier);
            $chantier->setBien($this);
        }

        return $this;
    }

    public function removeChantier(Chantier $chantier): static
    {
        if ($this->chantiers->removeElement($chantier)) {
            // set the owning side to null (unless already changed)
            if ($chantier->getBien() === $this) {
                $chantier->setBien(null);
            }
        }

        return $this;
    }

    /**
     * Get full address for display
     */
    public function getFullAddress(): string
    {
        return $this->adresseBien . ', ' . $this->villeBien;
    }

    public function __toString(): string
    {
        return $this->getFullAddress();
    }
}
